airflow/utils/orm_event_handlers.py

```diff
@@ -52,6 +55,40 @@ def set_mysql_timezone(dbapi_connection, connection_record):
             cursor.execute("SET time_zone = '+00:00'")
             cursor.close()
 
+        is_mariadb = engine.dialect.server_version_info and "MariaDB" in engine.dialect.server_version_info
+        if idle_session_timeout > 0:
+            if is_mariadb:
+                # https://mariadb.com/kb/en/mariadb-1032-release-notes/#variables
+                if engine.dialect.server_version_info > (10, 3):
+                    setting = "idle_write_transaction_timeout"
+                else:
+                    setting = "idle_readwrite_transaction_timeout"
+
+                @event.listens_for(engine, "connect")
+                def set_mysql_transaction_idle_timeout(dbapi_connection, connection_record):
+                    cursor = dbapi_connection.cursor()
+                    cursor.execute("SET @@SESSION.%s = %s", (setting, int(idle_session_timeout),))
+                    cursor.close()
+            else:
+                # MySQL doesn't have the equivalent of postgres' idle_in_transaction_session_timeout -- the
+                # best we can do is set the time we wait just set the timeout for any idle connection
+                @event.listens_for(engine, "connect")
+                def set_mysql_transaction_idle_timeout(dbapi_connection, connection_record):
```

Comment:
Do you think (from observation) it might have an undesireable effect for  MySQL ? From what I understand, the side effect of this change is that when workers will be idle, they will close and re-open connections continuously and frequently. From what I know this is much more problematic for Postgres (because closing/reopening the connection spawns a new process and is a costly event. 
For Postgres this is mitigated by PBBouncer (and with Postgres' transaction idle timeout, this will not happen, really (it's only some long, hanging transaction that will be impacted).

For MySQL also I understand that such wait timeout will close any connection that has no activity rather quickly (including those that have open transaction)? 

Is HA introducing some long hanging transactions in the long term that could impact this? Are you aware of some long transactions already present in Airflow?

Since the default value for MySQL is 28.880 (8 hours), I believe this one might have a tremendous impact on the behavior of MySQL in some cases. 

What is the effect on HA if we do not set this timeout to a long value?  It's not yet visible as we see only this PR for now but I would love to understand what kind of problems we can expect.

Reply:
Yeah, I was of two minds on this whole PR. As I say up above, this is not about performance of HA, but "what happens if a transaction goes wild and hangs on to a lock -- how long should we wait."

(HA PR coming -- I am doing the work to pull out as much as possible on to separate PRs as that one is already going to be huge enough)

The default for both postgres and mysql(likes) is the locking process/connection can wait For Ever (I think mysql's default is 1 year. I'm not kidding).

10s is perhaps a slow default, and 5mins/1 hour might be a better default, certainly until we have the time to do more testing of impact of this on mysql.

Reply:
> 10s is perhaps a low default, and 5mins/1 hour might be a better default, certainly until we have the time to do more testing of impact of this on mysql.

Have you tested HA with mysql in large scale?

Reply:
Yeah. We will definitely have to make some tests on MySQL. 

I am also a bit afraid about performance impact of that one for overall performance when the database is remote and there is an SSL enabled for example - the effect of this change might be that pretty much every transaction will start paying  the handshake penalty. This is probably neglectible for postgres/pgbouncer but for MySQL and Cloud SQL in particular, it might become a real problem. For now I think setting the default to original value if no HA should be OK I think and I'd love to see the locking mechanisms in HA scheduler later and see how big of an impact it will have.

Reply:
I have done some testing on MySQL, but not enough (and not with SSL) to answer this question.

I will come back to this PR later (hopefully pre 2.0, but if that is looking not possible I will change this to have default off), as this isn't on the critical path for AIP-15

Reply:
> > 10s is perhaps a low default, and 5mins/1 hour might be a better default, certainly until we have the time to do more testing of impact of this on mysql.
> 
> Have you tested HA with mysql in large scale?

Largest test I did was with 30 workers, 4 schedulers and 100k TIs to run. Not sure if that counts as large or not.

Reply:
Though thinking about it -- it likely wouldn't have hit this case very much.

Reply:
I think the only way to properly test it is to engage our users with some test/staging systems where they can simulate and test with production-like loads. I think the nature of this makes it super difficult to test. The problem is that the low-level behaviour that we have no way to control from the "higher-layer" logic which depends on factors that are pretty much random (i.e. how many transactions we run in parallel, whether some other threads were using the connections recently and returned it to the pool and how much load on the database we have in general (how responsive the DB server is).

I can easily imagine scenarios where  - for example - some workers will get periodic bursts of timed-out connections for all their threads when other workers start to do a lot of database transactions - because those workers will **just** cross the timeout limit and all of them start failing. Or a periodic failure of random tasks without a reason we could track or mitigate. Those are the most difficult things to track.

I do not have a very concrete scenario that could be easily simulated - It's more from intuition and experiencing similar problems in telecom industry years ago. We can try to mitigate it for example by adding some randomness around the timeout values and other things. but I think there is a good reason why the default timeout is 8 hrs and not say - 2 minutes. That makes me a little worried.

We might ask our customers to try this out once the whole thing is merged (including HA Scheduler). But I think during the process I would like to understand how critical it is to have this timeout. Maybe we can try some other approaches if all that we want to protect from is avoiding stale locks. Seems to me this is a bit risky to potentially change behaviour of all transactions when we want to protect from something that will happen very rarely.

For example, I could easily imagine we could run a periodic monitor to check if there are stale locks and selectively kill only those sessions.? That might call for a solution which is much more database-specific, but it might be less-risky, more efficient approach.